### PR TITLE
Functional tests for subscription triggering

### PR DESCRIPTION
**Pull request by Christoffer-Cortes**

### CODE REVIEW

File: src/functionaltests/java/com/ericsson/ei/subscriptions/trigger/SubscriptionTriggerSteps.java [hunk 1]
```diff
@@ -1,44 +1,312 @@
...
+    public void subscriptions_is_setup_using_REST_API(String endPoint) {
+        String readFileToString = "";
+        try {
+            readFileToString = FileUtils.readFileToString(new File(SUBSCRIPTION_WITH_JSON_PATH), "UTF-8");
+        } catch (IOException e) {
+            LOGGER.error(e.getMessage(), e);
+        }
+        readFileToString = stringReplaceText(readFileToString);
+        
+        ArrayList<String> subscriptions = new ArrayList<String>();
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/subscriptions/trigger/SubscriptionTriggerSteps.java]: should be List< String > subscriptions = new ArrayList<>();

File: src/functionaltests/java/com/ericsson/ei/subscriptions/trigger/SubscriptionTriggerSteps.java [hunk 2]
```diff
@@ -1,44 +1,312 @@
...
+        return table.count();
+    }
+
+    private boolean waitForEventsToBeProcessed(int eventsCount) {
+        int maxTime = 30;
+        int counterTime = 0;
+        long processedEvents = 0;
+        while (processedEvents < eventsCount && counterTime < maxTime) {
+            processedEvents = countProcessedEvents(database, collection);
+            LOGGER.info("Have gotten: " + processedEvents + " out of: " + eventsCount);
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/subscriptions/trigger/SubscriptionTriggerSteps.java]: logger should be debug

File: src/main/java/com/ericsson/ei/subscriptionhandler/InformSubscription.java [hunk 3]
```diff
@@ -164,6 +180,39 @@ public void informSubscriber(String aggregatedObject, JsonNode subscriptionJson)
...
+        String URL = splittednotificationMeta[0];
+        
+        List<NameValuePair> params = null;
+        try {
+            params = URLEncodedUtils.parse(new URI(notificationMeta), Charset.forName("UTF-8"));
+        } catch (Exception e) {
+            LOGGER.error(e.getMessage(), e);
+        }
+        
+        List<NameValuePair> processedParams = new ArrayList<NameValuePair>();
```
Comment (ned29) [inline src/main/java/com/ericsson/ei/subscriptionhandler/InformSubscription.java]: diamonds type List< NameValuePair > processedParams = new ArrayList<>();

File: src/main/java/com/ericsson/ei/subscriptionhandler/RunSubscription.java [hunk 4]
```diff
@@ -107,19 +107,25 @@ public boolean runSubscriptionOnObject(String aggregatedObject, Iterator<JsonNod
...
-                LOGGER.info("New Rule after replacing single quote : " + new_Rule);
+                LOGGER.debug("Rule : " + rule);
+                LOGGER.debug("New Rule after replacing single quote : " + new_Rule);
                 JsonNode result = jmespath.runRuleOnEvent(rule, aggregatedObject);
-                LOGGER.info("Result : " + result.toString());
-                if (result.toString() != null && result.toString() != "false" && !result.toString().equals("[]")) {
+                LOGGER.debug("Result : " + result.toString());
+                Boolean conclusion1 = !result.toString().equals("null");
+                Boolean conclusion2 = !result.toString().equals("false");
+                Boolean conclusion3 = !result.toString().equals("[]");
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/subscriptionhandler/RunSubscription.java]: I feel these Booleans can be renamed to conditions, just as in the logging output. More consistent and less confusing.

File: src/functionaltests/java/com/ericsson/ei/subscriptions/trigger/SubscriptionTriggerSteps.java [hunk 5]
```diff
@@ -1,44 +1,312 @@
...
+            try {
+                TimeUnit.MILLISECONDS.sleep(3000);
+                counterTime += 3;
+            } catch (InterruptedException e) {
+                LOGGER.error(e.getMessage(), e);
+            }
+        }
+        if(processedEvents == eventsCount) {
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/subscriptions/trigger/SubscriptionTriggerSteps.java]: simplify, just replace if condition with 
return processedEvents == eventsCount;

File: src/main/java/com/ericsson/ei/subscriptionhandler/InformSubscription.java [hunk 6]
```diff
@@ -94,12 +102,13 @@ public void informSubscriber(String aggregatedObject, JsonNode subscriptionJson)
...
         String notificationType = getSubscriptionField("notificationType", subscriptionJson);
         String notificationMeta = getSubscriptionField("notificationMeta", subscriptionJson);
-
         String key = "";
         String val = "";
         MultiValueMap<String, String> mapNotificationMessage = new LinkedMultiValueMap<>();
         ArrayNode arrNode = (ArrayNode) subscriptionJson.get("notificationMessageKeyValues");
-        if (arrNode.isArray()) {
+        
+        Boolean notificationMessageKeyValuesHasValues =  !arrNode.toString().equals("[]");
```
Comment (ned29) [inline src/main/java/com/ericsson/ei/subscriptionhandler/InformSubscription.java]: why Boolean? maybe should be boolean

File: src/functionaltests/java/com/ericsson/ei/subscriptions/trigger/SubscriptionTriggerSteps.java [hunk 7]
```diff
@@ -1,44 +1,308 @@
...
+    public void subscriptions_is_setup_using_REST_API(String endPoint) {
+        String readFileToString = "";
+        try {
+            readFileToString = FileUtils.readFileToString(new File(SUBSCRIPTION_WITH_JSON_PATH), "UTF-8");
+        } catch (IOException e) {
+            LOGGER.error(e.getMessage(), e);
+        }
+        readFileToString = stringReplaceText(readFileToString);
+        
+        ArrayList<String> subscriptions = new ArrayList<>();
```
Comment (ValentinTyhonov) [inline src/functionaltests/java/com/ericsson/ei/subscriptions/trigger/SubscriptionTriggerSteps.java]: Is there some reason why you did noy changed it to List? 

In PR title on Github we don't set JIRA story number.

File: src/main/java/com/ericsson/ei/subscriptionhandler/InformSubscription.java [hunk 8]
```diff
@@ -94,12 +102,13 @@ public void informSubscriber(String aggregatedObject, JsonNode subscriptionJson)
...
         String notificationMeta = getSubscriptionField("notificationMeta", subscriptionJson);
-
         String key = "";
         String val = "";
         MultiValueMap<String, String> mapNotificationMessage = new LinkedMultiValueMap<>();
         ArrayNode arrNode = (ArrayNode) subscriptionJson.get("notificationMessageKeyValues");
-        if (arrNode.isArray()) {
+
+        boolean notificationMessageKeyValuesHasValues = !arrNode.toString().equals("[]");
+        if (notificationMessageKeyValuesHasValues) {
```
Comment (ValentinTyhonov) [inline src/main/java/com/ericsson/ei/subscriptionhandler/InformSubscription.java]: if  you use this notificationMessageKeyValuesHasValues only one time, use !arrNode.toString().equals("[]") directly in if statement without creating variable.

File: src/main/java/com/ericsson/ei/subscriptionhandler/InformSubscription.java [hunk 9]
```diff
@@ -166,8 +180,43 @@ public void informSubscriber(String aggregatedObject, JsonNode subscriptionJson)
...
+     * This method reformats notificationMeta and fetches values if applicable
+     * for the requested parameters.
+     *
+     * @param aggregatedObject
+     * @param notificationMeta
+     * @return String
+     */
+    private String reformatNotificationMeta(String aggregatedObject, String notificationMeta) {
+        String[] splittednotificationMeta = notificationMeta.split("\\?");
+        String URL = splittednotificationMeta[0];
```
Comment (ValentinTyhonov) [inline src/main/java/com/ericsson/ei/subscriptionhandler/InformSubscription.java]: no need to create two variables here, if you can do next:
String URL = notificationMeta.split("\\?")[0];

File: src/main/java/com/ericsson/ei/jsonmerge/MergePrepare.java [hunk 10]
```diff
@@ -449,7 +449,9 @@ private int getOriginObjectArraySize(String originObject, JSONArray mergePathArr
             for (int i = 0; i < mergePathIndex; i++) {
                 String key = mergePathArray.getString(i);
                 if (valueForKey == null) {
-                    valueForKey = originJSONObject.get(key);
+                    if (originJSONObject.has(key)) {
+                        valueForKey = originJSONObject.get(key);
+                    }
                 } else {
                     if (valueForKey instanceof JSONObject) {
```
Comment (ValentinTyhonov) [inline src/main/java/com/ericsson/ei/jsonmerge/MergePrepare.java]: make else if statement here

File: src/main/java/com/ericsson/ei/jsonmerge/MergePrepare.java [hunk 11]
```diff
@@ -449,7 +449,9 @@ private int getOriginObjectArraySize(String originObject, JSONArray mergePathArr
             for (int i = 0; i < mergePathIndex; i++) {
                 String key = mergePathArray.getString(i);
                 if (valueForKey == null) {
-                    valueForKey = originJSONObject.get(key);
+                    if (originJSONObject.has(key)) {
```
Comment (ValentinTyhonov) [inline src/main/java/com/ericsson/ei/jsonmerge/MergePrepare.java]: you can use && instead of two if statements.

**Pull request merged**